[CORE-455] Add endpoint to update the billing account of a workspace

samanehsan opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CORE-455

**Testing:**
- Created two billing projects and a workspace in a BEE
- Called the Rawls endpoint to update the workspace billing
- Went to the workspace google project and confirmed changes to the billing account, project IAM and bucket IAM are correct
- Reloaded the workspace in the UI and confirmed that I got an error page because the workspace no longer exists under the original namespace
- Went to the workspaces page and confirmed that I see the workspace with the new namespace
---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [x] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

samanehsan committed:
Add update workspace billing endpoint

samanehsan committed:
Add endpoint validation

samanehsan committed:
Check that dest billing is enabled

samanehsan committed:
WIP - refactor and implement more of the billing changeover

samanehsan committed:
Check that new billing account is open and update IAM roles

samanehsan committed:
Revert billing change on error

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-455-update-workspace-billing

samanehsan committed:
Fix unmarshalling request

samanehsan committed:
Fix WorkspaceRequestUpdateBilling model

samanehsan committed:
Fix MockGoogleServicesDAO

Will fix areServicesEnabled later

samanehsan committed:
Fix checking dest billing is different than source

samanehsan committed:
Scala fmt and streamline validation

samanehsan committed:
Fix checking if billing projects are GCP

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-455-update-workspace-billing

samanehsan commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -643,6 +643,229 @@ class WorkspaceService(
+            )
+          )
+        else Future.unit
+      sourceBillingProject <- getBillingProjectContext(sourceBillingProjectName)
```
It looks like this only gets GCP billing projects, so `requireBillingProjectIsGCP` does not get reached if either the source or destination billing is in Azure.

samanehsan committed:
Refactor updating workspace billing

Call validation method and then call
the update method separately for easier
testing.

samanehsan committed:
Add tests for validating billing update

samanehsan committed:
Return updated workspace, not the original

samanehsan committed:
Fix cleaning up billing update after error

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-455-update-workspace-billing

calypsomatic commented on core/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -6336,6 +6380,16 @@ components:
+        newBillingProjectName:
+          type: string
+          description: The new namespace (billing project) to use for this workspace
+            AD is set)
```
"AD is set"?  Am I supposed to know what that means?

samanehsan replied:
Oh I think that's a mistake, thanks for catching! AD is auth domain but it doesn't make sense here

calypsomatic commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -555,6 +559,26 @@ class HttpGoogleServicesDAO(val clientSecrets: GoogleClientSecrets,
+      }
+      response.getOpen.booleanValue()
+    } {
+      case gjre: GoogleJsonResponseException if gjre.getStatusCode / 100 == 4 => false // any 4xx error means no access
```
So if it's a 500 or something it'll rethrow the error?

samanehsan replied:
Hmm let me take a closer look at what this is doing 👀 I think we don't want the `if gjre.getStatusCode / 100 == 4 => false` at least since a permissions error doesn't necessarily mean that the billing account is closed/deactivated

samanehsan committed:
Fix tests

samanehsan committed:
Fix WorkspaceRequestUpdateBilling description

samanehsan committed:
Make Rawls SA update workspace policies

samanehsan committed:
Fix test workspace

samanehsan committed:
Fix changing GCS bucket IAM

If a workspace has NO auth domains,
change the billing project owner policy
email that is a writer on the bucket to the
new billing project owner policy email.

samanehsan committed:
Fix getBillingAccountEnabled error handling

samanehsan committed:
Update tests

samanehsan committed:
Fix tests again

davidangb commented on core/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -4272,6 +4272,50 @@ paths:
+        - $ref: '#/components/parameters/workspaceNamespacePathParam'
+        - $ref: '#/components/parameters/workspaceNamePathParam'
+      requestBody:
+        description: Request for new workspace
```
nit: description is incorrect

samanehsan replied:
Ah this is copy pasta 🍝

davidangb commented on core/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -4272,6 +4272,50 @@ paths:
+              schema:
+                $ref: '#/components/schemas/ErrorReport'
+        404:
+          description: Source workspace not found
```
nit: "Source" is misleading here

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -555,6 +559,31 @@ class HttpGoogleServicesDAO(val clientSecrets: GoogleClientSecrets,
+    } { case e: GoogleJsonResponseException =>
+      throw new RawlsExceptionWithErrorReport(
+        ErrorReport(
+          StatusCodes.InternalServerError,
```
Could this try to propagate the status code that Google threw instead of always returning a 500? See  https://github.com/broadinstitute/rawls/blob/7180e1d969529bf77affc982f19dc0efe63175ec/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpSamDAO.scala#L113 as an example.

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/util/BillingProjectSupport.scala:
```diff
@@ -67,6 +67,20 @@ trait BillingProjectSupport {
+        )
+      )
+    } else Future.successful(())
+
```
any reason this needs to be a `Future`? It looks totally synchronous.

samanehsan replied:
Oh good point! `gcsDAO.areServicesEnabled` is a boolean and I had to wrap it in a Future to call it from within the for block: https://github.com/broadinstitute/rawls/blob/475dabdd8d123acbd023b24a8b84f5410c927311/core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala#L738

So I could change it to look like that (unless there's a better way to do this in Scala)

davidangb replied:
If you switch from `<-` to `=` you don't need any Future wrapping:
```scala
servicesEnabled  = gcsDAO.areServicesEnabled(GoogleProject(workspace.googleProjectId.value), storageAPIs)
```

davidangb commented on core/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -4272,6 +4272,50 @@ paths:
+              $ref: '#/components/schemas/WorkspaceRequestUpdateBilling'
+        required: true
+      responses:
+        201:
```
The implementation code returns a 200; update this? (I do think 200 is more appropriate than 201)

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -643,6 +643,275 @@ class WorkspaceService(
+              ErrorReport(StatusCodes.NotFound, s"Workspace ${workspaceName.name} does not exist")
+            )
+          )
+      }
```
I think you could reduce some of the boilerplate by using `=` instead of `<-`, here and in a few other places in this method, something like this:
```suggestion
      workspaceOpt <- workspaceRepository.getWorkspace(workspaceName)
      workspace = workspaceOpt.getOrElse(
        throw RawlsExceptionWithErrorReport(
          ErrorReport(StatusCodes.NotFound, s"Workspace ${workspaceName.name} does not exist")
        )
      )
```

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -643,6 +643,275 @@ class WorkspaceService(
+        logger.warn(s"Failed to roll back SAM billing update for workspace ${workspace.briefName}", rollbackEx)
+      }
+
+  def updateWorkspaceBilling(workspace: Workspace, destBillingProject: RawlsBillingProject): Future[Option[Workspace]] =
```
optional, up to you: for readability and organization, consider making a new class to hold the update and rollback methods for GCP and Sam. If that new class accepted the workspace, new billing account name, old policy email, and new policy email as parameters to its constructor, then the update/rollback methods wouldn't need to accept any parameters and this `updateWorkspaceBilling` would probably read a lot easier.

samanehsan replied:
I like the suggestion! I am going to skip it though in the interest of time

samanehsan replied:
I like the suggestion! I am going to skip it though in the interest of time

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-455-update-workspace-billing

samanehsan committed:
Fix API docs

samanehsan committed:
Simplify validation logic

samanehsan committed:
Re-throw original google error

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-455-update-workspace-billing

samanehsan committed:
More test updates

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -555,6 +559,31 @@ class HttpGoogleServicesDAO(val clientSecrets: GoogleClientSecrets,
+      response.getOpen.booleanValue()
+    } { case e: GoogleJsonResponseException =>
+      val response = Option(e.getContent).getOrElse(e.getMessage)
+      val errorReport = Try(response.parseJson.convertTo[ErrorReport]).recover { case _: Throwable =>
```
I don't think a `GoogleJsonResponseException` will ever parse into an `ErrorReport` so that attempt will always fail and fall through to the recover; you could remove it. My apologies for not being clear - in my original comment I was only suggesting propagating the http status code.

samanehsan replied:
Oh I see!

calypsomatic commented on core/src/test/scala/org/broadinstitute/dsde/rawls/mock/MockSamDAO.scala:
```diff
@@ -177,7 +177,9 @@ class MockSamDAO(dataSource: SlickDataSource)(implicit executionContext: Executi
 
-  override def getUserArbitraryPetServiceAccountKey(userEmail: String): Future[String] = ???
+  override def getUserArbitraryPetServiceAccountKey(userEmail: String): Future[String] = Future.successful(
+    """{"client_email": "pet-110347448408766049948@broad-dsde-dev.iam.gserviceaccount.com", "client_id": "104493171545941951815"}"""
```
Is this real?  Should it be?

samanehsan replied:
I have no idea, but I can take a closer look and find out! I copied it over from the other mock methods 😅

samanehsan committed:
Remove incorrect mocking

samanehsan committed:
Use FastPass mock to test happy path

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -3962,4 +4004,405 @@ class WorkspaceServiceSpec
+      val workspace = Await.result(services.workspaceRepository.getWorkspace(workspaceName), Duration.Inf).get
+      val targetBilling = testData.testProject1
+
+//    val userSubjectId = WorkbenchUserId(userInfo.userSubjectId.value)
```
tiny tiny nit: delete instead of commenting out

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -3962,4 +4004,405 @@ class WorkspaceServiceSpec
+      )
+      verifySamUpdate(oldBillingProjectOwnerPolicyEmail, newBillingProjectOwnerPolicyEmail, services)
+      verify(services.mockFastPassService, atLeastOnce()).removeFastPassGrantsForWorkspace(workspace)
+      verify(services.mockFastPassService, atLeastOnce()).syncFastPassesForUserInWorkspace(workspace)
```
these should only run once, correct? Could you verify once instead of `atLeastOnce()`?

EDIT: I see use of `atLeastOnce()` in a bunch of other tests in this PR; update those too?

samanehsan replied:
The tests that fail and do a rollback will call these methods twice but with different arguments. I tried to refactor it to specifically verify what it was called with each time but couldn't quite get it working 😞

samanehsan replied:
I did change the verify methods in the happy path test though, since those do get called just once

davidangb reviewed (commented):
Just a few pretty minor comments, nice job getting that test passing!

samanehsan committed:
Code review comments

samanehsan committed:
Merge branch 'develop' into se/CORE-455-update-workspace-billing

samanehsan committed:
Merge branch 'develop' into se/CORE-455-update-workspace-billing
